### PR TITLE
Handle comments in the prefix

### LINKED ISSUE #2861: Query with a leading multi-line /* */ comment not recognized by vtgate
After recently upgrading vtgate we've started getting the following errors from vtgate (slightly edited):
```
vtgate: http://apa4.sjc1.square:31901/: unrecognized statement: /* criteria query */ select this_.id as id1_35_0_, this_.created_at as created_2_35_0_, this_.updated_at as updated_3_35_0_, ... more columns..., this_.name as name7_35_0_ from mytable this_
```

I think the issues is the leading comment `/* criteria query */`. I'm going to see if disabling those in Hibernate helps.

### PR DESCRIPTION
**Pull request by tirsen**
Fixes #2861

### CODE REVIEW

**Comment by tirsen** [comment]
We tested this version with a version using `strings.HasPrefix` and this version is less readable but significantly faster:

```
BenchmarkTrim1-8       100000000            14.0 ns/op
BenchmarkTrim2-8       100000000            10.7 ns/op
```

(Trim2 is this version.)

10 ns overhead seems acceptable to me.

File: go/vt/sqlparser/analyzer.go [hunk 1]
```diff
@@ -49,10 +49,29 @@ const (
 // textual comparison to identify the statement type.
 func Preview(sql string) int {
 	trimmed := strings.TrimFunc(sql, unicode.IsSpace)
+
+	// Strip comments if necessary
+	if len(trimmed) > 1 && trimmed[0] == '/' && trimmed[1] == '*' {
```
Comment (sougou) [inline go/vt/sqlparser/analyzer.go]: I think you can use strings.HasPrefix here?

File: go/vt/sqlparser/analyzer.go [hunk 2]
```diff
@@ -49,10 +49,29 @@ const (
 // textual comparison to identify the statement type.
 func Preview(sql string) int {
 	trimmed := strings.TrimFunc(sql, unicode.IsSpace)
+
+	// Strip comments if necessary
```
Comment (sougou) [inline go/vt/sqlparser/analyzer.go]: We might as well put this in a loop to skip all leading comments :).

File: go/vt/sqlparser/analyzer.go [hunk 3]
```diff
@@ -49,10 +49,29 @@ const (
...
+	}
+
+	if len(trimmed) > 1 && trimmed[0] == '-' && trimmed[1] == '-' {
+		index := strings.Index(trimmed, "\n")
+		if index != -1 {
+			trimmed = trimmed[index+1:]
+			trimmed = strings.TrimFunc(trimmed, unicode.IsSpace)
+		}
+	}
+
```
Comment (sougou) [inline go/vt/sqlparser/analyzer.go]: Can you move this to comments.go, along with SplitTrailingComments?

**Comment by sougou** [comment]
I should l2read your comments before commenting :). Please ignore my comments about HasPrefix.

File: go/vt/sqlparser/analyzer.go (hunk 1 above)
Comment (tirsen) [inline go/vt/sqlparser/analyzer.go]: Yeah, it is about 50% slower though.

**Comment by tirsen** [comment]
We didn't benchmark the actual comment stripping. Only the conditional. But it does mean you only pay the price if you actually have a comment.

I think even the stripping should be fast though. Go slices are awesome.

**Pull request merged**